File: 0146-lru-cache/0146-lru-cache.cpp
#include <iostream>
#include <unordered_map>
using namespace std;
class Node {
  public:
  Node* prev;
  Node* next;
  int key;
  int value;
  Node(int key, int value){
    this->key = key;
    this->value = value;
    this->prev = nullptr;
    this->next = nullptr;
  };
};

class LRUCache {
  public:
  Node* head;
  Node* tail;
  unordered_map<int, Node*> cache;
  int capacity;
  
      LRUCache(int capacity) {
        this->capacity = capacity;
        this->head = new Node(0,0);
        this->tail = new Node(0,0);

        this->head->next = tail;
        this->tail->prev = head;

          
      }
      
      int get(int key) {

        // check if key is in 
        
        if(cache.find(key) != cache.end()){
          Node* resnode = cache[key];

          remove(resnode);
          add(resnode);

          return cache[key]->value;

        }
        
        return -1;
      }
      
      void put(int key, int value) {
        // check size first:
        if (cache.find(key) != cache.end()){
            Node* lastnode = cache[key];
            remove(lastnode);
            cache.erase(lastnode->key);
            delete lastnode;
            
          }
        else if (cache.size() == capacity){
          // remove the last node
          Node* lastnode = tail->prev;
          remove(lastnode);
          cache.erase(lastnode->key);
          delete lastnode;
        }

          Node* new_node = new Node(key,value);
          add(new_node);
          cache[key] = new_node;
          return;

          
      }
      void add(Node* node){
        node->next = this->head->next;
        node->prev = this->head;
        
        this->head->next->prev = node;
        this->head->next = node;
        

      }
      void remove(Node* node){
        node->next->prev = node->prev;
        node->prev->next = node->next;
      }
  };
  
  /**
   * Your LRUCache object will be instantiated and called as such:
   * LRUCache* obj = new LRUCache(capacity);
   * int param_1 = obj->get(key);
   * obj->put(key,value);
   */

File: 0146-lru-cache/0146-lru-cache.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.next = None
        self.prev = None

class LRUCache:

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.size =0 
        self.head = Node(0,0)
        self.tail = Node(0,0)
        self.head.next = self.tail
        self.tail.prev = self.head
        

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        else:
            node = self.cache[key]
            self.remove(node)
            self.add(node)
            return node.value
    def put(self, key: int, value: int) -> None:

        # check if key is already inside
        if key in self.cache:
            self.cache[key].value = value
            self.remove(self.cache[key])
            self.add(self.cache[key])
            return 
        else:
            if self.size == self.capacity:
                last_node = self.tail.prev
                self.remove(last_node)
                del self.cache[last_node.key]
                self.size-=1
            
            node = Node(key,value)
            self.add(node)
            self.size+=1
            self.cache[key] = node
            return

    def add(self, node):
        node.next = self.head.next
        node.prev = self.head
        self.head.next.prev = node
        self.head.next = node
    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)